cdap-app-fabric/src/main/java/co/cask/cdap/security/TokenSecureStoreUpdater.java

```diff
@@ -231,8 +238,36 @@ public long getUpdateInterval() {
 
   @Override
   public SecureStore update(String application, RunId runId) {
+    try {
+      if (Constants.Service.MASTER_SERVICES.equals(application)) {
+        // don't impersonate for the master.services Twill app
+        return doUpdate();
+      }
+
+      ProgramId programId = TwillAppNaming.fromTwillAppName(application);
+      return impersonator.doAs(programId.getNamespaceId(), new Callable<SecureStore>() {
+        @Override
+        public SecureStore call() throws Exception {
+          return doUpdate();
+        }
+      });
+    } catch (Exception e) {
+      // it should already be a runtime exception anyways, since none of the methods in the above callable
+      // throw any checked exceptions
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Invoked when an update to secure store is needed, but without impersonation
+   */
+  public SecureStore doUpdate() {
```

Comment:
nit: a ``do*`` method generally is a private helper method. If this is being exposed outside, can we just call it ``update()``? In fact, it should probably be a method on the base class as well.